src/circuit.rs

```diff
@@ -1,46 +1,658 @@
+            let blind = {
+                let rcv = ScalarFixed::new(
+                    ecc_chip.clone(),
+                    layouter.namespace(|| "rcv"),
+                    self.rcv.as_ref().map(|rcv| **rcv),
+                )?;
+                let value_commit_r = OrchardFixedBasesFull::ValueCommitR(PhantomData);
+                let value_commit_r = FixedPoint::from_inner(ecc_chip.clone(), value_commit_r);
+
+                // [rcv] ValueCommitR
+                value_commit_r.mul(layouter.namespace(|| "[rcv] ValueCommitR"), &rcv)?
+            };
+
+            // [v_net] ValueCommitV + [rcv] ValueCommitR
+            commitment.add(layouter.namespace(|| "cv_net"), &blind)?
+        };
+
+        // Nullifier integrity
+        // TODO: constrain to equal public input nf_old
+        let nf_old = {
+            // nk_rho_old = poseidon_hash(nk, rho_old)
+            let nk_rho_old = {
+                let message = [nk, rho_old];
+
+                let poseidon_message = layouter.assign_region(
+                    || "load message",
+                    |mut region| {
+                        let mut message_word = |i: usize| {
+                            let value = message[i].value();
+                            let var = region.assign_advice(
+                                || format!("load message_{}", i),
+                                config.poseidon_config.state[i],
+                                0,
+                                || value.ok_or(plonk::Error::SynthesisError),
+                            )?;
+                            region.constrain_equal(&config.perm, var, message[i].cell())?;
+                            Ok(Word::<_, _, poseidon::OrchardNullifier, 3, 2> {
+                                inner: StateWord::new(var, value),
+                            })
+                        };
+
+                        Ok([message_word(0)?, message_word(1)?])
+                    },
+                )?;
+
+                let poseidon_hasher = PoseidonHash::init(
+                    config.poseidon_chip(),
+                    layouter.namespace(|| "Poseidon init"),
+                    ConstantLength::<2>,
+                )?;
+                let poseidon_output = poseidon_hasher.hash(
+                    layouter.namespace(|| "Poseidon hash (nk, rho_old)"),
+                    poseidon_message,
+                )?;
+                let poseidon_output: CellValue<pallas::Base> = poseidon_output.inner.into();
+                poseidon_output
+            };
+
+            // Add hash output to psi using standard PLONK
+            // `scalar` = poseidon_hash(nk, rho_old) + psi_old.
+            //
+            let scalar = {
+                let scalar_val = nk_rho_old
+                    .value()
+                    .zip(psi_old.value())
+                    .map(|(nk_rho_old, psi_old)| nk_rho_old + psi_old);
+                let scalar = self.load_private(
```

Comment:
It seems that this variable is not connected to anything? It should be returned from this block instead of `ScalarFixed`.

Reply:
Thanks! This required us to support multiplying a fixed base by a base field element. `scalar` there is a base field element (a result of an addition `mod p`). It is now directly used in

```rust
nullifier_k.mul_base_field_elem(
    layouter.namespace(|| "[poseidon_output + psi_old] NullifierK"),
    scalar,
)?
```